server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistoryService.java

```diff
@@ -0,0 +1,119 @@
+
+    public MasterHistoryService(TransportService transportService, ThreadPool threadPool, ClusterService clusterService) {
+        this.transportService = transportService;
+        this.localMasterHistory = new MasterHistory(threadPool, clusterService);
+        this.clusterService = clusterService;
+    }
+
+    /**
+     * This returns the MasterHistory as seen from this node. The returned MasterHistory will be automatically updated whenever the
+     * ClusterState on this node is updated with new information about the master.
+     * @return The MasterHistory from this node's point of view. This MasterHistory object will be updated whenever the ClusterState changes
+     */
+    public MasterHistory getLocalMasterHistory() {
+        return localMasterHistory;
+    }
+
+    /**
+     * This method returns a static view of the MasterHistory on the node given. This MasterHistory is static in that it will not be
+     * updated even if the ClusterState is updated on this node or the remote node. The history is retrieved asynchronously, and only if
+     * requestRemoteMasterHistory has been called for this node. If anything has gone wrong fetching it, or if it has not been fetched in
+     * time for the current request, the returned value will be null.
+     * @param node The node whose MasterHistory we want
+     * @return The MasterHistory from the remote node's point of view. This MasterHistory object will not be updated with future changes
+     */
+    @Nullable
+    public List<DiscoveryNode> getRemoteMasterHistory(DiscoveryNode node) {
+        return nodeToHistoryMap.get(node);
+    }
+
+    /**
+     * This method attempts to fetch the master history from the requested node. If we are able to successfully fetch it, it will be
+     * available in a later call to getRemoteMasterHistory. The client is not notified if or when the remote history is successfully
+     * retrieved. This method only fetches the remote master history once, and it is never updated unless this method is called again.
+     * This is a remote call, so clients should avoid calling it any more often than necessary.
+     * @param node The node whose view of the master history we want to fetch
+     */
+    public void requestRemoteMasterHistory(DiscoveryNode node) {
+        long startTime = System.nanoTime();
+        transportService.openConnection(
+            // Note: This connection is explicitly closed onResponse or onFailure of the request below
+            node,
+            ConnectionProfile.buildDefaultConnectionProfile(clusterService.getSettings()),
+            new ActionListener<>() {
+                @Override
+                public void onResponse(Transport.Connection connection) {
+                    logger.trace("Opened connection to {}, making master history request", node);
+                    transportService.sendRequest(
+                        node,
+                        MasterHistoryAction.NAME,
```

Comment:
This action is new in 8.3(?) but we may be in the middle of a rolling upgrade in which `node` is an earlier version and does not understand this action. We should avoid sending this request at all if so.

Reply:
Done.